src/v/security/gssapi_authenticator.cc

```diff
@@ -0,0 +1,422 @@
+    gss::buffer output_token{};
+    OM_uint32 minor_status{};
+    auto major_status = gss_unwrap(
+      &minor_status, _context, &input_token, &output_token, nullptr, nullptr);
+    if (major_status != GSS_S_COMPLETE) {
+        fail(
+          major_status,
+          minor_status,
+          "gss {} failed to unwrap ssf of {} bytes",
+          _state,
+          auth_bytes.size());
+        return errc::invalid_credentials;
+    }
+
+    if (output_token.size() < 4) {
+        fail(
+          0,
+          0,
+          "gss {} unexpected data in unwrapped result of {} bytes",
+          _state,
+          output_token.size());
+    }
+
+    if (auto res = check(); res.has_error()) {
+        return res.assume_error();
+    }
+
+    bytes ret{};
+    vlog(seclog.trace, "gss {} sending {} bytes", _state, ret.size());
+    finish();
+    return ret;
+}
+
+result<void> gssapi_authenticator::check() {
+    OM_uint32 major_status{};
+    OM_uint32 minor_status{};
+    OM_uint32 lifetime_rec{};
+    gss::name source;
+    gss::name target;
+    gss_OID mech{};
+    OM_uint32 ctx_flags{};
+    int open{};
+
+    major_status = ::gss_inquire_context(
+      &minor_status,
+      _context,
+      &source,
+      &target,
+      &lifetime_rec,
+      &mech,
+      &ctx_flags,
+      nullptr,
+      &open);
+    if (major_status != GSS_S_COMPLETE) {
+        fail(
+          major_status,
+          minor_status,
+          "gss {} failed to inquire context",
+          _state);
+        return errc::invalid_scram_state;
+    }
+
+    auto target_buf = target.display_name_buffer();
+    std::string_view target_name{target_buf};
+    if (target_name.empty()) {
+        fail(0, 0, "gss {} failed to get service principal", _state);
+        return errc::invalid_scram_state;
+    }
+
+    auto source_buf = source.display_name_buffer();
+    std::string_view source_name{source_buf};
+    if (source_name.empty()) {
+        fail(0, 0, "gss {} failed to get client principal", _state);
+        return errc::invalid_scram_state;
+    }
+
+    vlog(
+      seclog.debug,
+      "gss_inquire_context: source: {}, target: {}, lifetime_rec: {}, "
+      "ctx_flags: {}, open: {}",
+      source_name,
+      target_name,
+      lifetime_rec,
+      display_ctx_flags(ctx_flags),
+      open);
+
+    // None of this is needed - left here temporarily for posterity
+    // if ((ctx_flags & GSS_C_INTEG_FLAG) == 0) {
+    //     // No integrity
+    //     _state = state::failed;
+    //     return errc::invalid_credentials;
+    // } else if ((ctx_flags & GSS_C_CONF_FLAG) == 0) {
+    //     // No confidentiality
+    //     _state = state::failed;
+    //     return errc::invalid_credentials;
+    // }
+
+    _principal = acl_principal{principal_type::user, ss::sstring{source_name}};
+    return outcome::success();
+}
+
+void gssapi_authenticator::finish() {
+    _context.reset();
+    _server_creds.reset();
+    _state = state::complete;
+}
```

Comment:
Is there like a re-roll that's going to happen?  Or is that a TBD thing to handle?  Is that part of the 'ticket' that kerberos produces?